include/hibf/hierarchical_interleaved_bloom_filter.hpp

```diff
@@ -373,4 +407,118 @@ hierarchical_interleaved_bloom_filter::membership_agent() const
+    hierarchical_interleaved_bloom_filter const * const hibf_ptr{nullptr};
+
+    //!\brief Helper for recursive bulk counting.
+    template <std::ranges::forward_range value_range_t>
+    void bulk_count_impl(value_range_t && values, int64_t const ibf_idx, size_t const threshold)
+    {
+        auto agent = hibf_ptr->ibf_vector[ibf_idx].template counting_agent<value_t>();
+        auto & result = agent.bulk_count(values);
+
+        value_t sum{};
+
+        for (size_t bin{}; bin < result.size(); ++bin)
+        {
+            sum += result[bin];
+            auto const current_filename_index = hibf_ptr->ibf_bin_to_user_bin_id[ibf_idx][bin];
+
+            if (current_filename_index < 0) // merged bin
+            {
+                if (sum >= threshold)
+                    bulk_count_impl(values, hibf_ptr->next_ibf_id[ibf_idx][bin], threshold);
+                sum = 0u;
+            }
+            else if (bin + 1u == result.size() ||                                                  // last bin
+                     current_filename_index != hibf_ptr->ibf_bin_to_user_bin_id[ibf_idx][bin + 1]) // end of split bin
+            {
+                if (sum >= threshold)
+                    result_buffer[current_filename_index] = sum;
+                sum = 0u;
+            }
+        }
+    }
+
+    //!\brief Stores the result of bulk_count().
+    counting_vector<value_t> result_buffer;
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    counting_agent_type() = default;                                       //!< Defaulted.
+    counting_agent_type(counting_agent_type const &) = default;            //!< Defaulted.
+    counting_agent_type & operator=(counting_agent_type const &) = delete; //!< Deleted. hibf_ptr is const.
+    counting_agent_type(counting_agent_type &&) = default;                 //!< Defaulted.
+    counting_agent_type & operator=(counting_agent_type &&) = delete;      //!< Deleted. hibf_ptr is const.
+    ~counting_agent_type() = default;                                      //!< Defaulted.
+
+    /*!\brief Construct a counting_agent_type for an existing hierarchical_interleaved_bloom_filter.
+     * \private
+     * \param hibf The hierarchical_interleaved_bloom_filter.
+     */
+    explicit counting_agent_type(hierarchical_interleaved_bloom_filter const & hibf) :
+        hibf_ptr(std::addressof(hibf)),
+        result_buffer(hibf_ptr->number_of_user_bins)
+    {}
+    //!\}
+
+    /*!\name Counting
+     * \{
+     */
+    /*!\brief Counts the occurrences in each bin for all values in a range.
+     * \tparam value_range_t The type of the range of values. Must model std::ranges::forward_range. The reference type
+     *                       must model std::unsigned_integral.
+     * \param[in] values The range of values to process.
+     * \param[in] threshold Do not recurse into merged bins with less than this many hits. Default: 1.
```

Comment:
Mentioning merged bins is too technical IMHO

```suggestion
     * \param[in] threshold Only report counts that exceed the threshold. Range: [1,inf]. Default: 1.
```